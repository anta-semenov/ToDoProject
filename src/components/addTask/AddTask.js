import React, { Component } from 'react'
import PureRenderMixin from 'react-addons-pure-render-mixin'
import { NEW_TASK_TITLE } from '../../constants/defaults'
import './AddTask.less'

export default class AddTask extends Component {
  state = {value: ''}
  constructor(props) {
    super(props)
    this.shouldComponentUpdate = PureRenderMixin.shouldComponentUpdate.bind(this)
  }

  handleKeyDown = e => {
    switch (e.keyCode) {
      case 13: {
        this.props.addTask(this.refs.input.value)
        this.refs.input.value = ''
      }
    }
  }
  handleButtonClick = () => {
    this.props.addTask(this.refs.input.value)
    this.refs.input.value = ''
  }

  render() {
    return (
      <div className='add-task'>
        <input type='text' ref='input' className='add-task__textfield' onKeyDown={this.handleKeyDown} placeholder={NEW_TASK_TITLE}/>
        <button className='add-task__button' onClick={this.handleButtonClick} >Add Task</button>
      </div>
    )
  }
}

AddTask.propTypes = {
  addTask: React.PropTypes.func.isRequired
}
