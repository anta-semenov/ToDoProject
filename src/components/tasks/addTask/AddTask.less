@import '../../../styles/variables.less';
@import '../../../styles/commonStyles.less';

.add-task {
  position: relative;
  z-index: 1;
  margin: 0 0 35px 0;
  max-width: @size-max-width-tasks;
  width: 100%;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  flex-shrink: 0;
  border: 1px solid @color-border-gray;
  border-radius: 2px;

  &::after {
    content: '';
    position: absolute;
    z-index: -1;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    opacity: 0;
    border-radius: 2px;
    box-shadow: 0 0 4px fade(@color-accent, 40%);
    transition: opacity @transition-basic;
  }

  &:hover {
    border-color: fade(@color-accent, 50%);
  }

  &.has-focus {
    border-color: @color-accent;
    &::after {
      opacity: 1;
    }
  }
  &.is-empty {
    max-width: @size-max-width-tasks - 100;
  }
}

.add-task__textfield {
  position: relative;
  width: 100%;
  margin: 0 0px 0 0;
  padding: 10px 10px;
  font-size: @font-size-basic;
  font-weight: 400;

  //border: 1px solid @color-border-light-gray;
  border: none;
  background-color: transparent;
  border-top-left-radius: 2px;
  border-bottom-left-radius: 2px;
  transition: @transition-basic;

  &:hover {
  }
  &:focus,
  &:active {
    background-color: transparent;
    outline: none;
  }
  &::-webkit-input-placeholder {
    .placeholder();
  }
}

.add-task__button {
  position: relative;
  margin: 0;
  padding: 10px 20px;
  white-space: nowrap;
  flex-shrink: 0;

  font-size: @font-size-basic;
  font-weight: @font-weight-medium;
  color: @color-font-light;
  background-color: transparent;
  border: none;
  outline: none;
  border-left: 1px solid @color-border-light-gray;
  cursor: pointer;
  transition: border-color @transition-basic, color @transition-basic, background-color @transition-basic;

  .add-task.has-focus & {
    border-color: @color-border-light-gray;
    color: @color-accent;
    //background-color: fade(desaturate(@color-accent, 50%), 10%);
  }
  &:focus {
    outline: none;
  }
  .add-task.has-focus &:hover {
    background-color: fade(desaturate(@color-accent, 35%), 5%);
    border-color: @color-accent;
    color: @color-accent;
  }
}
.add-task__button-sign {
  position: relative;
  margin-left: 5px;
  vertical-align: baseline;
  font-size: 16px;

}
