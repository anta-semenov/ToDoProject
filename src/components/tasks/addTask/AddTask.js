import React, { Component } from 'react'
import { NEW_TASK_TITLE } from '../../../constants/defaults'
import classNames from 'classnames'
import './AddTask.less'

export default class AddTask extends Component {

  handleKeyDown = e => {
    switch (e.keyCode) {
      case 13: {
        this.props.addTask(this.refs.input.value)
        this.input.value = ''
      }
    }
  }
  handleFocus = () => {
    this.frame.classList.add('has-focus')
  }
  handleBlur = () => {
    this.frame.classList.remove('has-focus')
  }
  handleButtonClick = () => {
    this.props.addTask(this.input.value)
    this.input.value = ''
    this.input.focus()
  }

  render() {
    const inputClasses = classNames({
      'add-task__textfield': true,
      'is-empty': this.props.isSectionEmpty
    })
    return (
      <div className='add-task' ref={(c) => this.frame = c}>
        <input
          type='text'
          ref={(c) => this.input = c}
          className={inputClasses}
          onKeyDown={this.handleKeyDown}
          onFocus={() => this.handleFocus()}
          onBlur={() => this.handleBlur()}
          placeholder={NEW_TASK_TITLE} />
        <button className='add-task__button' onClick={this.handleButtonClick} ><span className='add-task__button-text'>Add Task</span></button>
      </div>
    )
  }
}

AddTask.propTypes = {
  addTask: React.PropTypes.func.isRequired,
  isSectionEmpty: React.PropTypes.bool
}
AddTask.defaultProps = {
  isSectionEmpty: false
}
