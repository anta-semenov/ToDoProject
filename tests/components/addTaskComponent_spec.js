import React from 'react'
import { mount } from 'enzyme'

import AddTask from '../../src/components/tasks/addTask/AddTask'

beforeEach(() => {
  jest.clearAllMocks()
})

describe('AddTask component tests', () => {
  it('Should render textfield and button', () => {
    const addTaskComponent = mount(<AddTask addTask={jest.fn()} setSearchQuery={jest.fn()} />)
    expect(addTaskComponent).toMatchSnapshot()
  })
  it('Should invoke addTask callback and clear the AddTextfield value when button click event occur', () => {
    const addTask = jest.fn()
    const addTaskComponent = mount(<AddTask addTask={addTask} setSearchQuery={jest.fn()} />)

    addTaskComponent.find('.add-task__button-text').simulate('click')
    expect(addTask).toHaveBeenCalled()
  })
  it('Should invoke setSearchQuery callback and clear the AddTextfield value when button click event occur', () => {
    const addTask = jest.fn()
    const setSearchQuery = jest.fn()
    const addTaskComponent = mount(<AddTask addTask={addTask} setSearchQuery={setSearchQuery} />)

    addTaskComponent.find('input').simulate('change', { target: { value: 'add' } })
    expect(addTaskComponent).toMatchSnapshot()
    expect(setSearchQuery).toHaveBeenCalledWith('add')

    addTaskComponent.find('.add-task__button-text').simulate('click')
    expect(addTask).toHaveBeenCalled()
    expect(setSearchQuery).toHaveBeenCalledWith('')
  })

  it('Should invoke addTask callback and clear the AddTextfield value when enter is hitted', () => {
    const addTask = jest.fn()
    const addTaskComponent = mount(<AddTask addTask={addTask} setSearchQuery={jest.fn()} />)

    addTaskComponent.find('input').simulate('keyDown', { key: 'Enter', keyCode: 13, which: 13 })
    expect(addTask).toHaveBeenCalled()
  })
})
