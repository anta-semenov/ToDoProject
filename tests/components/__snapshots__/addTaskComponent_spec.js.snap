// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTask component tests Should invoke setSearchQuery callback and clear the AddTextfield value when button click event occur 1`] = `
<AddTask
  addTask={[MockFunction]}
  hasFocus={false}
  isSectionEmpty={false}
  setSearchQuery={
    [MockFunction] {
      "calls": Array [
        Array [
          "add",
        ],
      ],
    }
  }
>
  <div
    className="add-task"
  >
    <input
      className="add-task__textfield"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="New Task"
      type="text"
    />
    <button
      className="add-task__button"
      onClick={[Function]}
    >
      <span
        className="add-task__button-text"
      >
        Add Task
      </span>
    </button>
  </div>
</AddTask>
`;

exports[`AddTask component tests Should render textfield and button 1`] = `
<AddTask
  addTask={[MockFunction]}
  hasFocus={false}
  isSectionEmpty={false}
  setSearchQuery={[MockFunction]}
>
  <div
    className="add-task"
  >
    <input
      className="add-task__textfield"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="New Task"
      type="text"
    />
    <button
      className="add-task__button"
      onClick={[Function]}
    >
      <span
        className="add-task__button-text"
      >
        Add Task
      </span>
    </button>
  </div>
</AddTask>
`;
